examples/simple-game-server/main.go

```diff
@@ -0,0 +1,503 @@
+	shutdownDelay := flag.Int("automaticShutdownDelayMin", 0, "If greater than zero, automatically shut down the server this many minutes after the server becomes allocated")
+	udp := flag.Bool("udp", true, "Server will listen on UDP")
+	tcp := flag.Bool("tcp", false, "Server will listen on TCP")
+
+	flag.Parse()
+	if ep := os.Getenv("PORT"); ep != "" {
+		port = &ep
+	}
+	if epass := os.Getenv("PASSTHROUGH"); epass != "" {
+		p := strings.ToUpper(epass) == "TRUE"
+		passthrough = &p
+	}
+	if eready := os.Getenv("READY"); eready != "" {
+		r := strings.ToUpper(eready) == "TRUE"
+		readyOnStart = &r
+	}
+	if eudp := os.Getenv("UDP"); eudp != "" {
+		u := strings.ToUpper(eudp) == "TRUE"
+		udp = &u
+	}
+	if etcp := os.Getenv("TCP"); etcp != "" {
+		t := strings.ToUpper(etcp) == "TRUE"
+		tcp = &t
+	}
+
+	if !*udp && !*tcp {
+		log.Fatalf("No protocol enabled, exiting")
+	}
+
+	log.Print("Creating SDK instance")
+	s, err := sdk.NewSDK()
+	if err != nil {
+		log.Fatalf("Could not connect to sdk: %v", err)
+	}
+
+	log.Print("Starting Health Ping")
+	stop := make(chan struct{})
+	go doHealth(s, stop)
+
+	if *passthrough {
+		var gs *coresdk.GameServer
+		gs, err = s.GameServer()
+		if err != nil {
+			log.Fatalf("Could not get gameserver port details: %s", err)
+		}
+
+		p := strconv.FormatInt(int64(gs.Status.Ports[0].Port), 10)
+		port = &p
+	}
+
+	if *readyOnStart {
+		log.Print("Marking this server as ready")
+		ready(s)
+	}
+
+	if *shutdownDelay > 0 {
+		shutdownAfterAllocation(s, *shutdownDelay)
+	}
+
+	if *udp {
+		go udpListener(port, s, stop)
+	}
+
+	if *tcp {
+		tcpListener(port, s, stop)
+	}
+}
+
+// doSignal shutsdown on SIGTERM/SIGKILL
+func doSignal() {
+	stop := signals.NewStopChannel()
+	<-stop
+	log.Println("Exit signal received. Shutting down.")
+	os.Exit(0)
+}
+
+// shutdownAfterAllocation creates a callback to automatically shut down
+// the server a specified number of minutes after the server becomes
+// allocated.
+func shutdownAfterAllocation(s *sdk.SDK, shutdownDelay int) {
+	err := s.WatchGameServer(func(gs *coresdk.GameServer) {
+		if gs.Status.State == "Allocated" {
+			time.Sleep(time.Duration(shutdownDelay) * time.Minute)
+			shutdownErr := s.Shutdown()
+			if shutdownErr != nil {
+				log.Fatalf("Could not shutdown: %v", shutdownErr)
+			}
+		}
+	})
+	if err != nil {
+		log.Fatalf("Could not watch Game Server events, %v", err)
+	}
+}
+
+func udpListener(port *string, s *sdk.SDK, stop chan struct{}) {
+	log.Printf("Starting UDP server, listening on port %s", *port)
+	conn, err := net.ListenPacket("udp", ":"+*port)
+	if err != nil {
+		log.Fatalf("Could not start UDP server: %v", err)
+	}
+	defer conn.Close() // nolint: errcheck
+	udpReadWriteLoop(conn, stop, s)
+}
+
+func udpReadWriteLoop(conn net.PacketConn, stop chan struct{}, s *sdk.SDK) {
+	b := make([]byte, 1024)
+	for {
+		sender, txt := readPacket(conn, b)
+		parts := strings.Split(strings.TrimSpace(txt), " ")
+
+		switch parts[0] {
```

Comment:
Can we transform this into a function call, which will be used for both TCP and UDP handlers?
something like
```
func generateResponse(msg string, s *sdk.SDK, stop chan, err error) (response string , toContinue bool, err error)
```

Reply:
Agreed, will implement it. What about calling it `func handleResponse(...)`? Since it will also call other functions depending on the message sent.

Reply:
yes, seems reasonable

Reply:
Thanks I see that now we have one function for both UDP and TCP, that's great.